### PR TITLE
Spliting update_next_page method

### PR DESCRIPTION
**Pull request by alexbourret**

### CODE REVIEW

File: python-lib/pagination.py [hunk 1]
```diff
@@ -62,27 +67,49 @@ def set_counting_key(self, counting_key):
...
+        else:
+            batch_size = 1
+        if self.skip_key:
+            self.skip = data.get(self.skip_key)
+            logger.info("update_next_page_offset:skip=data[{}]={}".format(self.skip_key, self.skip))
+        self.records_to_skip = self.records_to_skip + batch_size
+        logger.info("update_next_page_offset:records_to_skip={}, batch_size={}".format(self.records_to_skip, batch_size))
+
+
+    def update_next_page_per_page(self, data, response_links=None):
```
Comment (liamlynch-data) [inline python-lib/pagination.py:101]: I think it *might* be safe/wise to factor out a method to calculate batch_size from both the per page and offset methods. That's make them easier to read too. But would need to be careful.

File: python-lib/pagination.py [hunk 2]
```diff
@@ -95,31 +122,41 @@ def update_next_page(self, data, response_links=None):
                 self.is_last_batch_empty = True
         else:
             batch_size = 1
+        if self.skip_key:
+            self.skip = data.get(self.skip_key)
+            logger.info("update_next_page_per_page:skip=data[{}]={}".format(self.skip_key, self.skip))
+        self.records_to_skip = self.records_to_skip + batch_size
```
Comment (liamlynch-data) [inline python-lib/pagination.py:128]: I'm not sure we need to keep track of `records_to_skip` for a page based method? AFAIK we only need to calculate `batch_size` to see if we are at the end?

**Review by liamlynch-data — COMMENTED** [review]
I think splitting the method out for the 3 types of pagination was an awesome idea - it makes it much clearer what is going on.

So it's definitely getting there, but I was still quite confused about a lot of things and I think it needs a bit of fine tuning. I have made various suggestions/observations. Obviously we can talk about it - I may well have misunderstood things.

Also as these changes impact all of the pagination methods I will need to do some more testing around it when the changes are made - so meanwhile I will think about that and have a look at the existing unit and integration scenarios.

File: python-lib/pagination.py [hunk 3]
```diff
@@ -95,31 +122,41 @@ def update_next_page(self, data, response_links=None):
...
-        if self.total:
-            self.remaining_records = self.total - self.records_to_skip
-            logger.info("update_next_page:remaining_records={}, total={}, records_to_skip={}".format(
-                self.remaining_records, self.total, self.records_to_skip
-            ))
+
+    def update_next_page_default(self, data, response_links=None):
+        self.is_first_batch = False
+        self.counter += 1
+        self.next_page_number = self.next_page_number + 1
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/pagination.py:159]: The default is used in cases where no pagination mode is selected. I implemented the "no pagination mode" that way so that I don't have to switch cases in various place of the code according to whether or not the pagination is on enabled...

[the remark this answers is not in the document]
Comment (liamlynch-data) [reply python-lib/pagination.py]: Yep that makes sense - but I'm not sure why the method needs to do anything in that case - how the self.counter or self.next_page_number  get used - but I'm probably missing something....

File: python-lib/pagination.py [hunk 4]
```diff
@@ -95,31 +122,41 @@ def update_next_page(self, data, response_links=None):
...
+        logger.info("update_next_page_per_page:records_to_skip={}, batch_size={}".format(self.records_to_skip, batch_size))
+
+
+    def update_next_page_link(self, data, response_links=None):
+        self.next_page_url = None
+        response_links = response_links or {}
+        next_link = response_links.get('next', {})
+        next_page_url = next_link.get("url")
+        self.is_first_batch = False
+        self.counter += 1
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/pagination.py:136]: Done in a7b3509

File: python-lib/pagination.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/pagination.py]: Regarding these 3 update next page methods, I started by duplicating them from the initial method, then remove things very conservatively, which might explain all the left overs... I cleaned most of it in a7b3509

File: python-lib/pagination.py [hunk 5]
```diff
@@ -11,9 +11,6 @@ def __init__(self, config=None, skip_key=None, limit_key=None, total_key=None, n
         self.next_page_key = None
         self.next_page_url_base = None
         self.skip_key = None
-        self.limit_key = None
-        self.total_key = None
-        self.total = None
         self.next_page_url = None
         self.remaining_records = None
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/pagination.py]: removed in a7b3509

File: python-lib/pagination.py [hunk 6]
```diff
@@ -95,31 +122,41 @@ def update_next_page(self, data, response_links=None):
...
         if self.next_page_key:
             next_page_path = extract_key_using_json_path(data, self.next_page_key)
             if self.next_page_url_base and next_page_path:
                 self.next_page_url = "/".join([self.next_page_url_base, next_page_path])
             else:
                 self.next_page_url = next_page_path
-            logger.info("update_next_page:next_page_url_base={}, next_page_path={}, next_page_url={}".format(
+            logger.info("update_next_page_link:next_page_url_base={}, next_page_path={}, next_page_url={}".format(
                 self.next_page_url_base, next_page_path, self.next_page_url
             ))
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/pagination.py]: done in a7b3509

File: python-lib/pagination.py [hunk 7]
```diff
@@ -62,27 +67,49 @@ def set_counting_key(self, counting_key):
...
+            if extracted_data:
+                batch_size = len(extracted_data)
+            else:
+                batch_size = 0
+                self.is_last_batch_empty = True
+        else:
+            batch_size = 1
+        if self.skip_key:
+            self.skip = data.get(self.skip_key)
+            logger.info("update_next_page_offset:skip=data[{}]={}".format(self.skip_key, self.skip))
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/pagination.py]: The skip key is used when building the querystring of the next page url, so I removed it from here in a7b3509

File: python-lib/pagination.py (hunk 1 above)
Comment (alexbourret) [reply python-lib/pagination.py:95]: Done in [3b46290](https://github.com/dataiku/dss-plugin-api-connect/pull/30/commits/3b46290e47409bd30b45e5a541d452a6fff32bde)

**Review by liamlynch-data — APPROVED** [review]
The code looks good now and I have done some pretty exhaustive testing around it (details in shortcut card).

So I'm approving, but it would be nice if @StanislasBertrand is able to take a look also.

**Review by StanislasBertrand — APPROVED** [review]
Overall code LGTM, It reads quite easily, methods have a single responsibility, clear name, the logic of configure_paging defining the update_next_page makes sense. 

Some ideas to make it potentially a bit clearer if you have the time some day : 
* There are 2 ways to initialise pagination, with a `config` and wit arguments to `configure_paging`. But in the plugin the config is never used so why keep it ? it makes the init more confusing.

* pagination has a lot of attributes which can make it hard to follow what does what. A lot of the attributes of pagination are actually not really pagination attributes the state of the current "page". For instance data_is_list, could maybe just be passed as argument to `has_next_page` ? Or maybe it could be interesting to have just one attribute of pagination that is `current_page_state` dict and that stores all of these kind of atttributes
